### PR TITLE
OGM-1468 Support merge and refresh loaders in OgmEntityPersister

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/OGM-1468

### CODE REVIEW

File: core/src/test/java/org/hibernate/ogm/test/cascade/CascadeEntityLoaderTest.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
+/*
+ * Hibernate OGM, Domain model persistence for NoSQL datastores
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.ogm.test.cascade;
```
Comment (DavideD) [inline core/src/test/java/org/hibernate/ogm/test/cascade/CascadeEntityLoaderTest.java]: Shouldn't this test run for all the datastores?

File: core/src/test/java/org/hibernate/ogm/utils/OgmTestCase.java [hunk 2]
```diff
@@ -123,4 +124,14 @@ public void inTransaction(Consumer<Session> consumer) {
 		}
 	}
 
+	protected <T> void deleteAll(Class<T> entity, Serializable... ids) {
```
Comment (DavideD) [inline core/src/test/java/org/hibernate/ogm/utils/OgmTestCase.java]: Can you just call this `delete`? I prefer to keep the suffix `All` only for cases where it's not possible to specify the ids.

File: core/src/test/java/org/hibernate/ogm/test/cascade/CascadeEntityLoaderTest.java (hunk 1 above)
Comment (fax4ever) [reply core/src/test/java/org/hibernate/ogm/test/cascade/CascadeEntityLoaderTest.java]: Yes, it should. I'm going to move the package.

File: core/src/main/java/org/hibernate/ogm/persister/impl/OgmEntityPersister.java [hunk 3]
```diff
@@ -909,20 +911,16 @@ protected void createLoaders() {
...
 				"merge",
-				createEntityLoader( LockMode.READ )
-				//new CascadeEntityLoader( this, CascadingAction.MERGE, getFactory() )
+				new CascadeEntityLoader( this, CascadingActions.MERGE, getFactory() )
 			);
 		loaders.put(
 				"refresh",
-				createEntityLoader( LockMode.READ )
-				//new CascadeEntityLoader( this, CascadingAction.REFRESH, getFactory() )
+				new CascadeEntityLoader( this, CascadingActions.REFRESH, getFactory() )
```
Comment (gsmet) [inline core/src/main/java/org/hibernate/ogm/persister/impl/OgmEntityPersister.java]: With these changes, I think we can remove the whole overriding `createLoaders()` method and use the one from ORM. That was my ultimate goal for this issue as I made significant changes to ORM in this area and I would like these changes to come for free to OGM.

Could you try and check?

File: core/src/test/java/org/hibernate/ogm/utils/OgmTestCase.java (hunk 2 above)
Comment (fax4ever) [reply core/src/test/java/org/hibernate/ogm/utils/OgmTestCase.java]: delete was my first choice too. Unfortunately many implementations of *OGMBaseTest* already use that name :(

File: core/src/test/java/org/hibernate/ogm/test/cascade/CascadeEntityLoaderTest.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+/**
+ * Testing merge and refresh {@link CascadeEntityLoader}.
+ * Those loader coming in action on subsequent Entity load from data store,
+ * when merging detached Entity or refreshing attached Entity,
+ * performing operation on the Entity itself or from Parent Entity enabling cascade.
+ *
```
Comment (gsmet) [inline core/src/test/java/org/hibernate/ogm/test/cascade/CascadeEntityLoaderTest.java]: This comment is very unclear. I'm not even sure to understand what you mean. Could you maybe try to make shorter sentences?

File: core/src/main/java/org/hibernate/ogm/persister/impl/OgmEntityPersister.java (hunk 3 above)
Comment (fax4ever) [reply core/src/main/java/org/hibernate/ogm/persister/impl/OgmEntityPersister.java]: Very good idea. Remove code it is always better ;) I'm going to check if it really possible

File: core/src/test/java/org/hibernate/ogm/test/cascade/CascadeEntityLoaderTest.java (hunk 4 above)
Comment (fax4ever) [reply core/src/test/java/org/hibernate/ogm/test/cascade/CascadeEntityLoaderTest.java]: You're right. I'm going to cut and rephrase it.

**Comment by fax4ever** [comment]
Thank you @DavideD and @gsmet. I'm going to apply the change you've proposed.

File: core/src/test/java/org/hibernate/ogm/utils/OgmTestCase.java (hunk 2 above)
Comment (fax4ever) [reply core/src/test/java/org/hibernate/ogm/utils/OgmTestCase.java]: WDYT about deleteEntities?

Comment (DavideD) [reply core/src/test/java/org/hibernate/ogm/utils/OgmTestCase.java]: No need to change it, leave it `deleteAll`

Comment (fax4ever) [reply core/src/test/java/org/hibernate/ogm/utils/OgmTestCase.java]: ok

File: core/src/main/java/org/hibernate/ogm/persister/impl/OgmEntityPersister.java (hunk 3 above)
Comment (fax4ever) [reply core/src/main/java/org/hibernate/ogm/persister/impl/OgmEntityPersister.java]: @gsmet currently is it not possible to remove the whole method. If we wanted this, we should resize the scope of the issue, adding more changes on it, or alternatively we could open a follow up issue.

**Comment by fax4ever** [comment]
I rebased it, with some changes.

File: core/src/main/java/org/hibernate/ogm/persister/impl/OgmEntityPersister.java (hunk 3 above)
Comment (gsmet) [reply core/src/main/java/org/hibernate/ogm/persister/impl/OgmEntityPersister.java]: What's the issue exactly? You're too vague to convince me :).

Comment (fax4ever) [reply core/src/main/java/org/hibernate/ogm/persister/impl/OgmEntityPersister.java]: :) nice approach. In the current implementation of Hibernate ORM, *noneLockLoader* and r*eadLockLoader* are moved out from the loaders map. So in order to use the whole base method we have to change the getAppropriateLoader too, leading us out of scope of OGM-1376, from what I think.

Comment (gsmet) [reply core/src/main/java/org/hibernate/ogm/persister/impl/OgmEntityPersister.java]: Hum, yes, you're right, `getLoaderByLockMode()` should have been accessible to the child classes. Bummer.

I opened https://hibernate.atlassian.net/browse/HHH-12615 to change that.

**Comment by gsmet** [comment]
@fax4ever you can merge it.

We'll remove the method once https://hibernate.atlassian.net/browse/HHH-12615 is fixed.

File: core/src/main/java/org/hibernate/ogm/persister/impl/OgmEntityPersister.java (hunk 3 above)
Comment (fax4ever) [reply core/src/main/java/org/hibernate/ogm/persister/impl/OgmEntityPersister.java]: nice followup!

**Comment by fax4ever** [comment]
Merged! Thank you guys for review

**Pull request closed without merging**